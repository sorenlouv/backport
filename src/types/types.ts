import { AxiosError, AxiosResponse } from 'axios';

/*
 * Github
 */

export interface GithubQuery {
  access_token: string;
  per_page: number;
  author?: string;
}

export interface GithubIssue {
  html_url: string;
  number: number;
}

export interface GithubCommit {
  commit: {
    message: string;
  };
  sha: string;
}

export interface GithubSearch<T> {
  items: T[];
}

export interface GithubPullRequestPayload {
  title: string;
  head: string;
  base: string;
  body?: string;
  maintainer_can_modify?: boolean;
}

// TODO: Make PR to DefinitelyTypes to make AxiosError a generic that takes the error response as T
export interface GithubApiError extends AxiosError {
  response?: AxiosResponse<{
    message: string;
    errors?: Array<{}>;
    documentation_url: string;
  }>;
}

/*
 * Config
 */

interface ConfigOptions {
  all?: boolean;
  multiple?: boolean;
  multipleCommits?: boolean;
  multipleBranches?: boolean;
}

export interface ProjectConfigRaw extends ConfigOptions {
  upstream: string;
  branches: Array<string | BranchChoice>;
  labels?: string[];
}

export interface ProjectConfig extends ConfigOptions {
  upstream: string;
  branchChoices: BranchChoice[];
  labels?: string[];
}

export interface GlobalConfig extends ConfigOptions {
  username: string;
  accessToken: string;
}

export type CombinedConfig = Partial<GlobalConfig> & Partial<ProjectConfig>;

/*
 * Options
 */

export interface BackportOptions {
  accessToken: string;
  all: boolean;
  labels: string[];
  multiple: boolean;
  multipleBranches: boolean;
  multipleCommits: boolean;
  upstream: string;
  username: string;
  branchChoices?: BranchChoice[];
  branches?: string[];
  sha?: string;
}

/*
 * PullRequest
 */

export interface PullRequest {
  html_url: string;
  number: number;
}

/*
 * Commit
 */

export interface Commit {
  sha: string;
  message: string;
  pullRequest?: number;
}

/*
 * Branch
 */

export interface BranchChoice {
  name: string;
  checked?: boolean;
}
