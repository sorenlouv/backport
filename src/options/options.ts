import { fetchDefaultRepoBranchAndPerformStartupChecks } from '../services/github/v4/fetchDefaultRepoBranchAndPerformStartupChecks';
import { PromiseReturnType } from '../types/PromiseReturnType';
import { setLogLevel, setRedactedAccessToken } from './../services/logger';
import { ConfigOptions } from './ConfigOptions';
import { getOptionsFromCliArgs } from './cliArgs';
import { getOptionsFromConfigFiles } from './config/config';
import { getValidatedOptions } from './getValidatedOptions';

export type BackportOptions = Readonly<PromiseReturnType<typeof getOptions>>;
export async function getOptions(
  argv: string[],
  optionsFromModule?: ConfigOptions
) {
  const optionsFromConfig = await getOptionsFromConfigFiles(
    optionsFromModule?.ci
  );
  const optionsFromCli = getOptionsFromCliArgs(
    { ...optionsFromConfig, ...optionsFromModule },
    argv
  );
  setRedactedAccessToken(optionsFromCli.accessToken);

  // set log level when all config options have been taken into account
  setLogLevel({ verbose: optionsFromCli.verbose });

  // TODO: move `getValidatedOptions` to `getOptionsFromCliArgs`
  const validatedOptions = getValidatedOptions(optionsFromCli);

  // TODO: make `username` optional by defaulting to `currentUsername`
  const { defaultBranch } = await fetchDefaultRepoBranchAndPerformStartupChecks(
    validatedOptions
  );

  return {
    ...validatedOptions,

    // use the default branch as source branch (normally master) unless an explicit `sourceBranch` has been given
    sourceBranch: validatedOptions.sourceBranch
      ? validatedOptions.sourceBranch
      : defaultBranch,
  };
}
