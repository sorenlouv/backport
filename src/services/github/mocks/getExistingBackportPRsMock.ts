export const getExistingBackportPRsMock = (repoName: string) => ({
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        repository: {
                          name: repoName,
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
          ],
        },
      },
    },
  },
});
