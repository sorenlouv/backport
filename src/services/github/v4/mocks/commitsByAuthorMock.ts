import { DataResponse } from '../fetchCommitsByAuthor';

export const commitsWithPullRequestsMock: DataResponse = {
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '2e63475c483f7844b0f2833bc57fdee32095bacb',
                message: 'Add ðŸ‘»',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                message: 'Add witch (#85)',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        baseRefName: 'master',
                        labels: {
                          nodes: [{ name: 'my-label-b' }],
                        },
                        mergeCommit: {
                          oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
                          message: 'Add witch (#85)',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 85,
                        timelineItems: {
                          edges: [],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        labels: {
                          nodes: [{ name: 'my-label-a' }],
                        },
                        mergeCommit: {
                          oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                          message:
                            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                        },
                        repository: {
                          name: 'kibana',
                          owner: {
                            login: 'elastic',
                          },
                        },
                        baseRefName: 'master',
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  title: 'some title',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                oid: '3827bbbaf39914eda4f02f6940189844375fd097',
                message: 'Add backport config',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
            {
              node: {
                oid: '5ea0da550ac191029459289d67f99ad7d310812b',
                message: 'Initial commit',
                associatedPullRequests: {
                  edges: [],
                },
              },
            },
          ],
        },
      },
    },
  },
};
