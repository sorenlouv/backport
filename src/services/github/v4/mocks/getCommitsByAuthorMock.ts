import { CommitByAuthorResponse } from '../fetchCommitsByAuthor';

export const getCommitsByAuthorMock = (
  repoName: string
): CommitByAuthorResponse => ({
  repository: {
    ref: {
      target: {
        history: {
          edges: [
            {
              node: {
                oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                message:
                  'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                associatedPullRequests: {
                  edges: [
                    {
                      node: {
                        baseRefName: 'master',
                        labels: {
                          nodes: [{ name: 'my-label-a' }],
                        },
                        mergeCommit: {
                          oid: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
                          message:
                            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                        },
                        repository: {
                          name: repoName,
                          owner: {
                            login: 'elastic',
                          },
                        },
                        number: 80,
                        timelineItems: {
                          edges: [
                            {
                              node: {
                                source: {
                                  __typename: 'PullRequest',
                                  number: 99,
                                  title: 'some title',
                                  state: 'MERGED',
                                  baseRefName: '6.3',
                                  commits: {
                                    edges: [
                                      {
                                        node: {
                                          commit: {
                                            message:
                                              'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
              },
            },
          ],
        },
      },
    },
  },
});
