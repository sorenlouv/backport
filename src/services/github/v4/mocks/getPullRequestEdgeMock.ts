import { PullRequestEdge } from '../fetchCommitsByAuthor';

export function getPullRequestEdgeMock({
  pullRequestNumber,
  timelinePullRequest,
}: {
  pullRequestNumber: number;
  timelinePullRequest: {
    title: string;
    commits: string[];
  };
}): PullRequestEdge {
  return {
    node: {
      labels: {
        nodes: [{ name: 'my-label-a' }],
      },
      mergeCommit: {
        oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
      },
      repository: {
        name: 'kibana',
        owner: {
          login: 'elastic',
        },
      },
      number: pullRequestNumber,
      timelineItems: {
        edges: [
          {
            node: {
              source: {
                __typename: 'PullRequest',
                title: timelinePullRequest.title,
                state: 'MERGED' as const,
                commits: {
                  edges: timelinePullRequest.commits.map((message) => ({
                    node: { commit: { message } },
                  })),
                },
                baseRefName: '7.x',
              },
            },
          },
        ],
      },
    },
  };
}
