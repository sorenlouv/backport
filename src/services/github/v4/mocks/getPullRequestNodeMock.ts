import { PullRequestNode } from '../getExistingTargetPullRequests';

export function getPullRequestNodeMock({
  pullRequestNumber,
  timelinePullRequest,
}: {
  pullRequestNumber: number;
  timelinePullRequest: {
    title: string;
    commits: string[];
  };
}): PullRequestNode {
  return {
    labels: {
      nodes: [{ name: 'my-label-a' }],
    },
    mergeCommit: {
      oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
      message: 'my commit message...1',
    },
    repository: {
      name: 'kibana',
      owner: {
        login: 'elastic',
      },
    },
    baseRefName: 'master',
    number: pullRequestNumber,
    timelineItems: {
      edges: [
        {
          node: {
            source: {
              __typename: 'PullRequest',
              title: timelinePullRequest.title,
              state: 'MERGED' as const,
              commits: {
                edges: timelinePullRequest.commits.map((message) => ({
                  node: { commit: { message } },
                })),
              },
              baseRefName: '7.x',
            },
          },
        },
      ],
    },
  };
}
