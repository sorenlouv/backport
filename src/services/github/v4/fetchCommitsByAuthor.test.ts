import nock from 'nock';
import { BackportOptions } from '../../../options/options';
import { mockGqlRequest } from '../../../test/nockHelpers';
import { CommitSelected, CommitChoice } from '../../../types/Commit';
import {
  fetchCommitsByAuthor,
  getExistingTargetPullRequests,
} from './fetchCommitsByAuthor';
import { commitsWithPullRequestsMock } from './mocks/commitsByAuthorMock';
import { getCommitsByAuthorMock } from './mocks/getCommitsByAuthorMock';
import { getPullRequestEdgeMock } from './mocks/getPullRequestEdgeMock';

const defaultOptions = {
  repoOwner: 'elastic',
  repoName: 'kibana',
  sourceBranch: 'master',
  accessToken: 'myAccessToken',
  username: 'sqren',
  author: 'sqren',
  maxNumber: 10,
  githubApiBaseUrlV3: 'https://api.github.com',
  githubApiBaseUrlV4: 'http://localhost/graphql',
} as BackportOptions;

const authorIdMockData = { user: { id: 'myUserId' } } as const;

describe('fetchCommitsByAuthor', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => {
    nock.cleanAll();
  });

  describe('when commit has an associated pull request', () => {
    let res: CommitSelected[];
    let authorIdCalls: ReturnType<typeof mockGqlRequest>;
    let commitsByAuthorCalls: ReturnType<typeof mockGqlRequest>;

    beforeEach(async () => {
      authorIdCalls = mockGqlRequest({
        name: 'AuthorId',
        statusCode: 200,
        body: { data: authorIdMockData },
      });

      commitsByAuthorCalls = mockGqlRequest({
        name: 'CommitsByAuthor',
        statusCode: 200,
        body: { data: commitsWithPullRequestsMock },
      });

      res = await fetchCommitsByAuthor(defaultOptions);
    });

    it('Should return a list of commits with pullNumber and existing backports', () => {
      const expectedCommits: CommitChoice[] = [
        {
          sha: '2e63475c483f7844b0f2833bc57fdee32095bacb',
          formattedMessage: 'Add ðŸ‘» (2e63475c)',
          originalMessage: 'Add ðŸ‘»',
          existingTargetPullRequests: [],
          targetBranchesFromLabels: [],
          sourceBranch: 'master',
        },
        {
          sha: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
          formattedMessage: 'Add witch (#85)',
          originalMessage: 'Add witch (#85)',
          pullNumber: 85,
          existingTargetPullRequests: [],
          targetBranchesFromLabels: [],
          sourceBranch: 'master',
        },
        {
          sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
          formattedMessage: 'Add SF mention (#80)',
          originalMessage:
            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
          pullNumber: 80,
          existingTargetPullRequests: [{ branch: '6.3', state: 'MERGED' }],
          targetBranchesFromLabels: [],
          sourceBranch: 'master',
        },
        {
          sha: '3827bbbaf39914eda4f02f6940189844375fd097',
          formattedMessage: 'Add backport config (3827bbba)',
          originalMessage: 'Add backport config',
          existingTargetPullRequests: [],
          targetBranchesFromLabels: [],
          sourceBranch: 'master',
        },
        {
          sha: '5ea0da550ac191029459289d67f99ad7d310812b',
          formattedMessage: 'Initial commit (5ea0da55)',
          originalMessage: 'Initial commit',
          existingTargetPullRequests: [],
          targetBranchesFromLabels: [],
          sourceBranch: 'master',
        },
      ];
      expect(res).toEqual(expectedCommits);
    });

    it('should call with correct args to fetch author id', () => {
      expect(authorIdCalls).toMatchSnapshot();
    });

    it('should call with correct args to fetch commits', () => {
      expect(commitsByAuthorCalls).toMatchSnapshot();
    });
  });

  describe('existingTargetPullRequests', () => {
    it('should return existingTargetPullRequests when repoNames match', async () => {
      const res = await getExistingBackportsByRepoName('kibana', 'kibana');
      const expectedCommits: CommitChoice[] = [
        {
          existingTargetPullRequests: [{ branch: '6.3', state: 'MERGED' }],
          formattedMessage: 'Add SF mention (#80)',
          originalMessage:
            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
          pullNumber: 80,
          sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
          sourceBranch: 'master',
          targetBranchesFromLabels: [],
        },
      ];
      expect(res).toEqual(expectedCommits);
    });

    it('should not return existingTargetPullRequests when repoNames does not match', async () => {
      const res = await getExistingBackportsByRepoName('kibana', 'kibana2');
      const expectedCommits: CommitChoice[] = [
        {
          existingTargetPullRequests: [],
          formattedMessage: 'Add SF mention (#80)',
          originalMessage:
            'Add SF mention (#80)\n\n* Add SF mention\r\n\r\n* Add several emojis!',
          pullNumber: 80,
          sha: '79cf18453ec32a4677009dcbab1c9c8c73fc14fe',
          sourceBranch: 'master',
          targetBranchesFromLabels: [],
        },
      ];
      expect(res).toEqual(expectedCommits);
    });
  });

  describe('when a custom github api hostname is supplied', () => {
    it('should be used in gql requests', async () => {
      const authorIdCalls = mockGqlRequest({
        name: 'AuthorId',
        statusCode: 200,
        body: { data: authorIdMockData },
        apiBaseUrl: 'http://localhost/my-custom-api',
      });

      const commitsByAuthorCalls = mockGqlRequest({
        name: 'CommitsByAuthor',
        statusCode: 200,
        body: { data: commitsWithPullRequestsMock },
        apiBaseUrl: 'http://localhost/my-custom-api',
      });

      await fetchCommitsByAuthor({
        ...defaultOptions,
        githubApiBaseUrlV4: 'http://localhost/my-custom-api',
      });

      expect(authorIdCalls.length).toBe(1);
      expect(commitsByAuthorCalls.length).toBe(1);
    });
  });
});

describe('getExistingTargetPullRequests', () => {
  afterEach(() => {
    nock.cleanAll();
  });

  it('should return a result when commit messages match', () => {
    const commitMessage = 'my message (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'a pr title',
        commits: ['my message (#1234)'],
      },
    });
    const existingPRs = getExistingTargetPullRequests(
      commitMessage,
      pullRequestEdge
    );
    expect(existingPRs).toEqual([{ branch: '7.x', state: 'MERGED' }]);
  });

  it('should not return a result when commit messages do not match', () => {
    const commitMessage = 'my message1 (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'a pr title',
        commits: ['my message2 (#1234)'],
      },
    });
    const existingPRs = getExistingTargetPullRequests(
      commitMessage,
      pullRequestEdge
    );
    expect(existingPRs).toEqual([]);
  });

  it('should return a result when commit message matches pull request title and number', () => {
    const commitMessage = 'my message (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'my message (#1234)',
        commits: ['the actual message'],
      },
    });
    const existingPRs = getExistingTargetPullRequests(
      commitMessage,
      pullRequestEdge
    );
    expect(existingPRs).toEqual([{ branch: '7.x', state: 'MERGED' }]);
  });

  it('should not return a result when only pull request title (and not pull number) matches', () => {
    const commitMessage = 'my message (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'my message (#1235)',
        commits: ['the actual message'],
      },
    });
    const existingPRs = getExistingTargetPullRequests(
      commitMessage,
      pullRequestEdge
    );
    expect(existingPRs).toEqual([]);
  });

  it('should return a result when first line of a multiline commit message matches', () => {
    const commitMessage = 'my message (#1234)';
    const pullRequestEdge = getPullRequestEdgeMock({
      pullRequestNumber: 1234,
      timelinePullRequest: {
        title: 'a pr title',
        commits: ['my message (#1234)\n\nsomething else'],
      },
    });
    const existingPRs = getExistingTargetPullRequests(
      commitMessage,
      pullRequestEdge
    );
    expect(existingPRs).toEqual([{ branch: '7.x', state: 'MERGED' }]);
  });
});

async function getExistingBackportsByRepoName(
  repoName1: string,
  repoName2: string
) {
  const commitsMock = getCommitsByAuthorMock(repoName1);

  mockGqlRequest({
    name: 'AuthorId',
    statusCode: 200,
    body: { data: authorIdMockData },
  });

  mockGqlRequest({
    name: 'CommitsByAuthor',
    statusCode: 200,
    body: { data: commitsMock },
  });

  return fetchCommitsByAuthor({
    ...defaultOptions,
    repoName: repoName2,
  });
}
