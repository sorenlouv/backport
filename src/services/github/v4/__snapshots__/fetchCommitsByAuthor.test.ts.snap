// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch author id 1`] = `
Array [
  Object {
    "query": "
    query AuthorId($login: String!) {
      user(login: $login) {
        id
      }
    }
  ",
    "variables": Object {
      "login": "sqren",
    },
  },
]
`;

exports[`fetchCommitsByAuthor when commit has an associated pull request should call with correct args to fetch commits 1`] = `
Array [
  Object {
    "query": "
    query CommitsByAuthor(
      $repoOwner: String!
      $repoName: String!
      $maxNumber: Int!
      $sourceBranch: String!
      $authorId: ID
      $historyPath: String
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        ref(qualifiedName: $sourceBranch) {
          target {
            ... on Commit {
              history(
                first: $maxNumber
                author: { id: $authorId }
                path: $historyPath
              ) {
                edges {
                  node {
                    oid
                    message
                    associatedPullRequests(first: 1) {
                      edges {
                        node {
                          ...ExistingTargetPullRequests
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }

    
  fragment ExistingTargetPullRequests on PullRequest {
    # Source PR
    number
    repository {
      name
      owner {
        login
      }
    }
    mergeCommit {
      oid
      message
    }
    labels(first: 50) {
      nodes {
        name
      }
    }
    baseRefName
    timelineItems(last: 20, itemTypes: CROSS_REFERENCED_EVENT) {
      edges {
        node {
          ... on CrossReferencedEvent {
            source {
              __typename

              # Target PRs
              ... on PullRequest {
                title
                state
                baseRefName
                commits(first: 20) {
                  edges {
                    node {
                      commit {
                        message
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }

  ",
    "variables": Object {
      "authorId": "myUserId",
      "historyPath": null,
      "maxNumber": 10,
      "repoName": "kibana",
      "repoOwner": "elastic",
      "sourceBranch": "master",
    },
  },
]
`;
